### ISSUE TITLE
Version 7 release

### ISSUE THREAD

**Issue by sungam3r** [graphql-dotnet/graphql-dotnet#3168]
I propose to release v6 along with server v7 after review all stuff for subscriptions.

https://github.com/graphql-dotnet/graphql-dotnet/milestone/22

**Comment by Shane32** [comment]
Sounds good.

I'd like to consider refactoring the complexity analyzer into a validation rule and removing all complexity-specific code from ExecutionOptions, etc -- possibly moving to a separate nuget package also.  Any thoughts?

As it relates to the above, there is now two different validation rules that rely on the analysis of fragments in regards to parent nodes.  Specifically, the complexity analyzer (if refactored into a validation rule), and the new server authorization rule.  Right now the complexity analyzer sorts the fragments and then runs an analysis on them, while the new authorization rule is written to queue results as they occur in the document, completing node analysis when results are available.  The latter was designed that way so that TypeInfo would be valid during processing (to minimize memory use), whereas the former creates an instance of TypeInfo during fragment analysis (wasting cpu cycles and memory).  I'm not sure if you have any ideas here, but perhaps it would be best if the node walker always walked the sorted fragments first, then the operation(s), to (1) optimize memory/cpu usage, (2) eliminate complexity from the rules, and (3) consolidate code.

Otherwise there's nothing I am looking to add to GraphQL.NET v6.  As for server v7, I just want to get through subscriptions and the new authorization rule, plus add support for ASP.NET MVC (a very trivial change), which I require.

What about code-first federation?  You had planned to review that PR as well.  I don't use federation of course.  And then you had talked about rewriting the schema printer, but the existing code is perfectly sufficient for my needs.  I'm not sure if you're still planning to do that.

lmk - thanks!

**Comment by sungam3r** [comment]
I would not move complexity into new package.

> whereas the former creates an instance of TypeInfo during fragment analysis

Where?

> What about code-first federation?

I am catastrophically lacking time. If not subscriptions/WebSockets, then maybe I would have been engaged in the federation rework. I postponed this work.

**Comment by SebastianStehle** [comment]
Can you not align the server and core version? It always confused me. Angular had the same problem and the just jumped from 2.0 to 4.0 If I remember correctly.

**Comment by sungam3r** [comment]
@Shane32 I'm fine to skip v6 and jump to v7 here but then we should maintain major version bumping in both projects.

**Comment by Shane32** [comment]
Sounds good to me

**Comment by Shane32** [comment]
> > whereas the former creates an instance of TypeInfo during fragment analysis
> 
> Where?

In the PR that was just merged

https://github.com/graphql-dotnet/graphql-dotnet/pull/3159/files#diff-f4a29656e8af8665e2803754180c77667b11b6eb6c3a11f11026fd45bdc7f0d5R20-R35

**Comment by Shane32** [comment]
I agree, which is why I brought it up. I can write a PR to refactor the complexity analyzer into a validation rule, if I have time.

**Comment by Shane32** [comment]
I guess I meant to say that `TypeInfo` is normally only created once while validating the entire document, whereas with the complexity analyzer it now is created twice - once for the complexity analyzer, and once for all the rest of the rules.

**Comment by sungam3r** [comment]
> I'd prefer that we move the complexity analyzer into a validation rule 

I want to do this for a long time, including for other reasons (more design than performance).

**Comment by Shane32** [comment]
I propose we set our goals for final release of GraphQL.NET v7 now.

My goals:
- [ ] #3198 / #3203 
- [ ] #3214 (necessary)
- [ ] #3200 (easy)
- [ ] Review migration notes

Goals for GraphQL.NET Server:
- [ ] https://github.com/graphql-dotnet/server/pull/849
- [ ] Review / rewrite documentation (see new docs I've already written [here](https://github.com/Shane32/GraphQL.AspNetCore3/blob/master/README.md))
- [ ] Add migration documentation
- [ ] Update UI nuget packages to support .NET 4.8

Do you have any goals @sungam3r ?

I think it will still be a few weeks before release, as #3203 needs some work (with time for review), and updating the documentation for the server project will take some time as well.

**Comment by sungam3r** [comment]
Now I have a seasonal rest from OSS, so to speak. There is a lot of things to do at home and there is no much desire and strength to take a new issues in work or even complete existing ones 😕 . Even code review may be difficult for me from time to time. The main pending issue for me - rewrite of schema printer (the last commit in working branch was February 12) and I'm 95% sure that it will flow into v8.

**Comment by sungam3r** [comment]
OK, I'll review tomorrow pending PRs from you. I suggest to take timeout for ~1 week after merging these last PRs. Maybe we will find something more worth to fix or something like that, who knows. Personally, I'll try to look through my issues/PRs and decide if I can finish work on them for v7. #3220 and #3221 IMO should be merged for v7.

**Comment by Shane32** [comment]
Updated todo list @sungam3r :
- [ ] #3222 (fix for v5)
- [ ] #3226
- [ ] https://github.com/graphql-dotnet/server/pull/866
- [ ] https://github.com/graphql-dotnet/server/pull/856

**Comment by Shane32** [comment]
@sungam3r Ok for me to release 7.0 now?

**Comment by sungam3r** [comment]
I want to make one more PR with changes of `Field` APIs.

**Comment by sungam3r** [comment]
I know.

**Comment by sungam3r** [comment]
Only #3245 left in v7 milestone.

**Comment by Shane32** [comment]
@sungam3r We need to release 7.0.  If we could put some concentrated effort towards discussing the outstanding PRs, we could release it tomorrow.  Most of these PRs have already been reviewed for the most part; we just need to make a final decision.  Some of these PRs can be moved into 7.1 as they are not breaking changes.

Please let me know what your thoughts are.  Here's what is outstanding:

| PR | Breaking | Notes |
|-|-|-|
| #3270 Add AutoRegisteringInterfaceGraphType | No[^1] | Good 'new feature' to release with 7.0; contains very little code that is actually new |
| #3284 Add Argument overload with CLR type | No | Complements #3247 (new Field overload with CLR type); simple change |
| #3276 ExposeExceptionDetails | No[^2] | As it deprecates an old member, it should be released with 7.0 with migration notes |
| #3273 / #3277 IConfigureExecution changes | Yes | Need to decide on this, or wait until 8.0 |
| #3272 Rewrite IDocumentCache | Yes | Waiting on #3273 / #3277 |

[^1]: Includes bug fixes necessary for the feature to function properly, but the bug fixes are not breaking changes.
[^2]: Deprecates old member

**Comment by sungam3r** [comment]
I updated #3276 and reviewed #3273/#3277 .

Regarding 
> As it deprecates an old member, it should be released with 7.0 with migration notes

I disagree. Obsoleting members is not a breaking change at least in that particular case. Nevertheless I can retarget PR to `develop`. Would you?

**Comment by sungam3r** [comment]
I plan to review other pending PRs today.

**Comment by sungam3r** [comment]
I reviewed #3270 . Also I want to review all PRs with assembly signing before v7 release.

**Comment by Shane32** [comment]
> I updated #3276 and reviewed #3273/#3277 .
> 
> Regarding
> 
> > As it deprecates an old member, it should be released with 7.0 with migration notes
> 
> I disagree. Obsoleting members is not a breaking change at least in that particular case. Nevertheless I can retarget PR to `develop`. Would you?

I just meant that it is an opportune time to do so, with permanent documentation describing the change. We do not typically release much documentation regarding interim releases. We also do not typically dwell on removing obsolete members from a new major version. Which means that releasing a feature like this within 5.4 or 7.1, while acceptable, does not have as good of a documentation trace as it would be in 7.0. 

In this case I suggest we target develop and include migration notes describing the new feature, but it is not mandatory.

Perhaps we should have within the migration document a section for members which have been marked as obsolete and not otherwise mentioned, such as AddMemoryCache or ExposeExceptionStackTrace.

**Comment by Shane32** [comment]
> I reviewed #3270 . Also I want to review all PRs with assembly signing before v7 release.

Sure

**Comment by Shane32** [comment]
> I reviewed #3270 . Also I want to review all PRs with assembly signing before v7 release.

Lmk when done.

**Comment by sungam3r** [comment]
I merged #3276 and in the review process of #3265. Should take some time to be sure that everything is OK.

**Comment by Shane32** [comment]
Shall I release Parser 8.1.0 and GraphQL.NET 7.0 and Server 7.0 tonight @sungam3r ?

**Comment by sungam3r** [comment]
Yep. It's time to do it. Also remember about https://github.com/graphql-dotnet/graphql-dotnet/issues/3238
